### ISSUE TITLE
Use readable values in API data instead of numbers - use Country: Nepal instead of Country: 123

### ISSUE THREAD

**Issue by gulfaraz** [IFRCGo/go-api#1143]
_I'm creating this feature request on behalf of @toffehoff who sent the details via email._

**Is your feature request related to a problem? Please describe.**
I’m trying to connect our systems with GO to automatically import 3W projects. What I see is that, for all pre-defined lists, I need to fill in the ID, which is very cumbersome.

Example: if I want to create a new entry in Project, I first need to find the IDs of the following attributes:
Programme Type
Sector
Operation Type
Status
Reporting National Society
Country
Event
Disaster Type
Regional Project
 
And maybe also the districts and tags (I haven’t tested that, since that is an array field)

Strangely, for the attribute Visibility I don’t need to do that….

**Describe the solution you'd like**
Instead of filling in the IDs, fill in the values and that the API process will search the corresponding IDs.
Similar to the Visibility-attribute 😊
To not break the current implementation, it could be implemented like: if the given value is a number, then assume it is an ID, if the given value isNaN, then search for the ID. 
 
…. And (since I haven’t tested that yet), I hope that, when a successful submission of a new record, in the return message the ID will be mentioned.

**Comment by gulfaraz** [comment]
The _solution_ is to accept values instead of their ids. Example: `operation_type: "PROGRAMME"` instead of `operation_type: 0` in the request body. This requires a rewrite of the existing clients (go-frontend) that requires significant effort.

A _workaround_ is to list possible values and their ids in a dictionary. Anyone who wishes to use the go-api should refer to this dictionary for the correct ids. To look up the ids for the `operation_type` property,
```
GET /api/v2/operation-types
{
    "operation_type": {
        "Programme": 0,
        "Emergency Operation": 1
    }
}
```

A _dirty workaround_ is to create a lookup endpoint that returns all values and their ids. This involves only one new endpoint as an ultimate lookup dictionary.
```
GET /api/v2/dictionary
{
    "operation_type": {
        "Programme": 0,
        "Emergency Operation": 1
    },
    "country": {
        "Nepal": 123,
        "Afghanistan": 14,
        "Angola": 18,
        ...
    },
    ...
}
```

I advise creating an HTML page in the API documentation where the dictionary information is available to anyone who cannot (or doesn't know how to) call the dictionary API.

**On discussion with @tovari, the preference is to implement the _dirty workaround_ to save dev effort.**

The current approach complicates features, limits potential and makes it difficult to understand and use the API. I recommend fixing the problem with the proposed _solution_ ASAP. It is an effort to rewrite the existing clients but I believe it's worth the effort for the long term success of the go-api.

References
1. [Stack Overflow - RESTful API design: list value options for query param](https://stackoverflow.com/questions/57691240/restful-api-design-list-value-options-for-query-param)
2. [Google API - ElevationStatus enum in Maps API](https://developers.google.com/maps/documentation/elevation/overview#ElevationStatus)
3. [Twitter API - Enums in Type column](https://developer.twitter.com/en/docs/twitter-api/tweets/search/api-reference/get-tweets-search-recent)
4. [Facebook API - privacy enum values in Group API](https://developers.facebook.com/docs/graph-api/reference/v11.0/group)

FYI @batpad @toffehoff

**Comment by batpad** [comment]
+cc @thenav56

**Comment by thenav56** [comment]
As @gulfaraz suggested we can create a separate endpoint to provide enum values mapping. But we shouldn't add **database models** there (like country), as they have their separate endpoint.
This will be the quickest solution and something we wanted to do to avoid storing static values separately in the client. Not sure if this is already done.

For accepting string value instead of enum (number here), one of the easiest ways will be to use **CharField** instead of **django_enum**. (Django provides [TextChoices](https://docs.djangoproject.com/en/3.2/ref/models/fields/#enumeration-types) in v3). [It's time to upgrade Django :smiley: ]



> Strangely, for the attribute Visibility I don’t need to do that….

The project's visibility field uses CharField instead of enum, that is why it's working.

**Comment by gulfaraz** [comment]
@thenav56 Thank you for the clear explanation.

If the go-frontend will benefit from this change, is it practical to do the CharField fix instead of the workaround?

Also, Is upgrading Django already on the roadmap? If not, should it be?

**Comment by batpad** [comment]
In general, I would much prefer if all our ENUMs were CharFields with short text that might map to a more descriptive string rather than the IntegerFields with ENUMs we currently use, so I definitely broadly agree with this and understand the frustration here.

I think moving over to Char ENUMs throughout the application will be a large refactor and risks breaking many things - I am not sure @thenav56 how you would feel about this just on 3w?

For now, I would definitely advocate for the "dirty" solution - create endpoints to map all the ENUMs or one consolidated endpoint that returns all ENUM values. For things that are models in the db, like countries, of course, we should use the existing endpoints as per @thenav56's comments above. For countries, specifically, I think it maybe nice to explore the idea of POST APIs accepting ISO codes instead of / in addition to IDs.

@thenav56 we do currently store a few ENUM values on the client (we moved anything that was coming from the db) - having those as lookups and the frontend fetch them dynamically would be great.

If it seems like the dirty solution is the way to go now, I'd suggest making a list of all the ENUM values we want the API to return mappings for, and ticket out creating those end-points? @thenav56 @gulfaraz do you have thoughts on whether this is better as a single endpoint that clubs together all the ENUMs used, or separate endpoints per ENUM? For me, separate seems cleaner, but it also seems potentially a bit painful for the client to make separate requests for each ENUM, so I'm on the fence.

> Also, Is upgrading Django already on the roadmap? If not, should it be?

It's definitely been well over a year since we last did a major upgrade, time passes :( - we should definitely put an upgrade to v3 on the roadmap, I'll create a ticket - although, I don't see that making a huge difference to our approach here currently.

**Comment by thenav56** [comment]
> I think moving over to Char ENUMs throughout the application will be a large refactor and risks breaking many things - I am not sure @thenav56 how you would feel about this just on 3w?

Yes, I forget about the data migration as well. Let's go with the dirty solution.

> If it seems like the dirty solution is the way to go now, I'd suggest making a list of all the ENUM values we want the API to return mappings for, and ticket out creating those end-points? @thenav56 @gulfaraz do you have thoughts on whether this is better as a single endpoint that clubs together all the ENUMs used, or separate endpoints per ENUM? For me, separate seems cleaner, but it also seems potentially a bit painful for the client to make separate requests for each ENUM, so I'm on the fence.

I think a single endpoint should be better (for both users and us), maybe add filters if the user needs limited enums?

**Comment by gulfaraz** [comment]
It's unfortunate that the Char ENUM refactor cannot happen but I understand its risk.

@batpad @thenav56 my vote is for the single endpoint

+1 to use ISO codes instead of country id

**Comment by nanometrenat** [comment]
One note re `Instead of filling in the IDs, fill in the values and that the API process will search the corresponding IDs.` - if considering doing this then need to make sure that special characters and multi-word values are also catered-for.  xref https://github.com/IFRCGo/go-api/issues/357#issuecomment-651715348 for previous slightly-related conversations.

Also, re ISO codes, need to bear in mind there's not always a 1:1 relationship between an ISO code and a location - some tables have multiple entries for one ISO code where multiple entities are in that location (e.g. RO/CO and NS) though I know this may have changed after all the recent geogeek amendments ;)

**Comment by toffehoff** [comment]
Hi all, can we at least use countrycodes instead of IDs for countries and National Societies. For most others I can then create a simple mapping on my end. Which is also a dirty solution.

I understand that changing the API from IDs to text will have a cascading effect on the frontend.
Hence, I suggested to make a "hybrid" solution. Meaning, filling in `operation_type: "PROGRAMME"` and  `operation_type: 0 ` would both be a valid requests. The API could then check whether the value submitted is a number, in that case treat it as an ID, if not, treat it as a value. With this an complete rewrite of the frontend would not be needed.

Regarding the ISO codes and the 1:1 relationship. That would currently already be a problem. The IDs are linked to country names. However, if you are referring to the difference between country and National Society, I have already suggested to the geogeek lead (Johnny) so make this in to two separate entities: A geographical area belongs to a country and is represented by a NS, those are two different entities.

BTW: I hope this discussion has also lead to a design standard (or updated one) for future API calls :-)

**Comment by nanometrenat** [comment]
@toffehoff @batpad and team does this need to stay open given the much-improved API documentation and enum availability?
